### PR TITLE
Fix the race between failing and starting pending streams.

### LINKED ISSUE #330: OkHttpClientTransport.onGoAway() races with startPendingStreams()
onGoAway has two phases: do things necessary under lock and final cleanup. In the first phase it collects the streams to terminate in the second and sets goAway.

startPendingStreams() does not observe goAway and also creates new streams that should be failed due to the goAway. From an initial look, it seems it would be best to remove failPendingStreams() and simply integrate its two phases into onGoAway()'s two phases; that is, when holding the lock in onGoAway, replace pendingStreams with an empty list, and then when not holding the lock call transportReportStatus

### PR DESCRIPTION
**Pull request by madongfly**
Fixes #330

### CODE REVIEW

File: okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java [hunk 1]
```diff
@@ -269,29 +269,18 @@ private void startStream(OkHttpClientStream stream, List<Header> requestHeaders)
   private boolean startPendingStreams() {
     boolean hasStreamStarted = false;
     synchronized (lock) {
-      while (!pendingStreams.isEmpty() && streams.size() < maxConcurrentStreams) {
-        PendingStream pendingStream = pendingStreams.poll();
-        startStream(pendingStream.clientStream, pendingStream.requestHeaders);
-        pendingStream.createdFuture.set(null);
-        hasStreamStarted = true;
+      if (!goAway) {
```
Comment (ejona86) [inline okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java]: This goAway check is unnecessary, as pendingStreams is guaranteed to be empty (newStream() checks goAway before adding to pendingStreams). If it is ever non-empty, then that is a separate issue and needs to be fixed, independent of whether we have the check here.

File: okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java [hunk 2]
```diff
@@ -382,6 +372,9 @@ private void onGoAway(int lastKnownStreamId, Status status) {
           }
         }
       }
+
+      pendingStreamsCopy = pendingStreams;
+      pendingStreams = null;
```
Comment (ejona86) [inline okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java]: Can we use Collections.emptyList() instead?

File: okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java (hunk 1 above)
Comment (madongfly) [inline okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java]: Done.

File: okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java (hunk 2 above)
Comment (madongfly) [inline okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java]: We can't, since we use pendingStreams as a Queue instead of a List.

**Comment by ejona86** [comment]
@madongfly, LGTM

**Comment by madongfly** [comment]
@ejona86 PTAL, thanks!

**Pull request merged**